Source/NSFileCoordinator.m

```diff
@@ -1,7 +1,245 @@
+/* Definition of class NSFileCoordinator
+   Copyright (C) 2019 Free Software Foundation, Inc.
+   
+   Implemented by: 	Gregory Casamento <greg.casamento@gmail.com>
+   Date: 	Sep 2019
+   Original File by: Daniel Ferreira
+
+   This file is part of the GNUstep Library.
+   
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+   
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Library General Public License for more details.
+   
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free
+   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02111 USA.
+*/
+
 #import <Foundation/NSFileCoordinator.h>
+#import <Foundation/NSURL.h>
+#import <Foundation/NSArray.h>
+#import <Foundation/NSDictionary.h>
+#import <Foundation/NSFilePresenter.h>
+#import <Foundation/NSOperation.h>
+
+static NSMutableArray *__presenters = nil;
+static NSMutableDictionary *__presenterMap = nil;
 
 @implementation NSFileAccessIntent
+- (instancetype) init
+{
+  self = [super init];
+  if(self != nil)
+    {
+      _url = nil;
+      _isRead = NO;
+      _options = 0L;
+    }
+  return self;
+}
+
++ (instancetype) readingIntentWithURL: (NSURL *)url
+                              options: (NSFileCoordinatorReadingOptions)options
+{
+  NSFileAccessIntent *result = [[self alloc] init];
+  ASSIGNCOPY(result->_url, url);
+  result->_options = options;
+  result->_isRead = YES;
+  return result;
+}
+
++ (instancetype) writingIntentWithURL: (NSURL *)url
+                              options: (NSFileCoordinatorWritingOptions)options
+{
+  NSFileAccessIntent *result = [[self alloc] init];
+  ASSIGNCOPY(result->_url, url);
+  result->_options = options;
+  result->_isRead = NO;
+  return result;
+}
+
+- (NSURL *) URL
+{
+  return _url;
+}
 @end
 
 @implementation NSFileCoordinator
+
++ (void) initialize
+{
+  if(self == [NSFileCoordinator class])
+    {
+      __presenters = [[NSMutableArray alloc] init];
+      __presenterMap = [[NSMutableDictionary alloc] init];
+    }
+}
+
++ (NSArray *) filePresenters
+{
+  return __presenters;
+}
+
++ (void) addFilePresenter: (id)presenter
+{
+  [__presenters addObject: presenter];
+  [__presenterMap setObject: presenter forKey: [presenter presentedItemURL]];
+}
+
++ (void) removeFilePresenter: (id)presenter
+{
+  [__presenters removeObject: presenter];
```

Comment:
Shouldn't you remove it from __presenterMap as well?

Reply:
Fixed.